### PR TITLE
Upgrade to SDK v3 (original / split into smaller PR's)

### PR DESCRIPTION
**Pull request by kurtmckee**
This upgrades the automate client to SDK v3. It's possible that this is best reviewed by clicking on the first commit and working through the changes commit by commit.

I don't think that this work is complete, and certainly not as tested as it needs to be, but I'd like to get feedback on the work so far. From the changelog fragment:

Features
--------

-   Upgrade to Globus SDK v3.
-   Add code linting, documentation build testing, and a bunch of unit tests.
    (Code coverage of modules currently imported during testing is 73%.)

Bugfixes
--------

-   Fix a bug that prevented the Flows client from properly validating flow definition states in lists.
-   Fix a bug that could allow the Flows authorizer to be lost if an exception was raised.
    (Authorizer swaps are now handled using a context manager.)
-   Prevent empty values from appearing in query parameters or JSON submissions.
-   Fix a bug that prevented the input schema of an existing Flow from being set to an all-permissive JSON schema.

### CODE REVIEW

File: changelog.d/20211029_162341_kurtmckee_sdk_v3.rst
Comment (rudyardrichter) [inline changelog.d/20211029_162341_kurtmckee_sdk_v3.rst:6]: I would characterize this as a different heading ("Changes" or "Other"? I don't know if these exist by default, at least in this project) since it's not user-facing

File: globus_automate_client/cli/auth.py
Comment (rudyardrichter) [inline globus_automate_client/cli/auth.py:84]: The annotation here is effectively (useful) documentation, does a linter complain about this?

File: globus_automate_client/queues_client.py
Comment (rudyardrichter) [inline globus_automate_client/queues_client.py:99]: If I'm reading this correctly, this line would be a no-op since `data` only has the key-value pair `"data": {"id": ...}`?

File: pyproject.toml
Comment (rudyardrichter) [inline pyproject.toml:89]: Given this is a pretty large chunk maybe better to have it in a `tox.ini`? That way you'd actually get syntax highlighting/checking on this too which might be nicer.

**Review by rudyardrichter — CHANGES_REQUESTED** [review]
I did not make a pass through the tests yet, but here are some early thoughts.

- Others may differ, but I would be in favor of squashing the commits down pretty aggressively. For instance, I might condense
```
- Resolve isort errors
- Resolve flake8 errors
- Resolve most mypy errors
```
to just one `Apply linters` or similar. Adding tests in particular I'd be in favor of squashing because it's not likely we'd want to `git bisect` those later or anything.

File: globus_automate_client/flows_client.py
Comment (rudyardrichter) [inline globus_automate_client/flows_client.py:968]: I think this should be reverted. As the docstring says, an empty list actually has a particular meaning

**Review by urimandujano — COMMENTED** [review]
This looks like a few different PRs worth of work combined into one. Can you split this PR into smaller chunks based on what the code changes are? If we go that route I think any bugfixes that get made would be good to merge and release as part of this package's final 0.12.3 release. Then we can bump our dependency on globus-sdk to 3 and call that our 0.13.0 release.

**Review by rudyardrichter — CHANGES_REQUESTED** [review]

**Comment by rudyardrichter** [comment]
I'd second Uriel, it seems like this PR has a lot going on and it would be easier to track the SDK-upgrade-specific changes if that was separate from linting changes and the like.

File: changelog.d/20211029_162341_kurtmckee_sdk_v3.rst
Comment (kurtmckee) [reply changelog.d/20211029_162341_kurtmckee_sdk_v3.rst:6]: Agreed. It's not difficult to add these, it just needs to be codified in the config. For the scriv config I only copied what I saw in past release announcements:

```
categories = ["Features", "Bugfixes", "Documentation"]
```

Adding a tag like "Other" seems good to me. @jimPruyne / @urimandujano?

File: globus_automate_client/cli/auth.py
Comment (kurtmckee) [reply globus_automate_client/cli/auth.py:84]: My memory is that mypy didn't complain but PyCharm highlighted this as extraneous. It's annotating a dictionary key's value instead of the dictionary itself so it didn't seem to add additional value. Should I revert this?

File: globus_automate_client/queues_client.py
Comment (kurtmckee) [reply globus_automate_client/queues_client.py:99]: 100% correct! I only wrote unit tests for flows and action client code so this will definitely need to be a part of the queues client tests.

File: pyproject.toml
Comment (kurtmckee) [reply pyproject.toml:89]: Good point, I'll do that!

File: globus_automate_client/cli/auth.py
Comment (rudyardrichter) [reply globus_automate_client/cli/auth.py:84]: I'd vote in favor of keeping because it made it easier to understand what `self.tokens` is doing.

File: globus_automate_client/flows_client.py
Comment (kurtmckee) [reply globus_automate_client/flows_client.py:968]: 100% correct. I was on a kick to remove empty values. This needs to be undone.

File: globus_automate_client/cli/auth.py
Comment (kurtmckee) [reply globus_automate_client/cli/auth.py:84]: I'll get this fixed.

Separately, the token cache annotations triggered a number of errors throughout the codebase with mypy. Eventually this needs to be addressed better than what I did in 36f84f1026483cbb7b6d249b8b50fde114eec5c8.

**Comment by kurtmckee** [comment]
@urimandujano and @rudyardrichter I'll work to squash the commits and stage them as separate PR's. Based on what I'm reading, the suggestion is to have separate PR's for:

* Fixing bugs (including removing empty/false-y values from query parameters and JSON data)
* Adding unit tests / linters
* Upgrading to SDK v3

They'll probably be in this order, too. I'll work to get this done.

File: changelog.d/20211029_162341_kurtmckee_sdk_v3.rst
Comment (kurtmckee) [reply changelog.d/20211029_162341_kurtmckee_sdk_v3.rst:6]: I've updated the changelog fragment but it's still worthwhile to consider the categories.

**Comment by kurtmckee** [comment]
I've force-pushed updates to the original branch in case any other comments are shared. I'm working on separating the test/fix/upgrade commits into separate PR's now.

**Comment by rudyardrichter** [comment]
Thanks @kurtmckee, I realize it's a bunch of extra git work to get those organized, and for what will likely become the same end result, so your efforts are appreciated.

**Comment by kurtmckee** [comment]
Bug fix PR is live at #98. I'd like to wait for that to get reviewed so I can build the unit testing on top of it.

**Comment by kurtmckee** [comment]
It turns out the unit tests are heavily dependent on SDK v3. There is a great deal of breakage if the unit tests come next so I'm going to prep the SDK v3 changes on top of the bug fixes first. The unit tests and linting will be the final PR.

File: changelog.d/20211029_162341_kurtmckee_sdk_v3.rst
Comment (urimandujano) [reply changelog.d/20211029_162341_kurtmckee_sdk_v3.rst]: I'm not sure what to call these -- maybe "Internal", "Development" or "Meta changes"? I don't have a strong opinion but I'm willing to go with whatever feels good to you two.

**Comment by urimandujano** [comment]
@kurtmckee I appreciate your extra effort on breaking these out. I took a look at #98 and that looks good to me. If we end up merging that then I think we could be good maintainers and cut a release there.

**Comment by jimPruyne** [comment]
What's the current state of this PR? Is this still a to be reviewed, or has refactors / splits / etc made this obsolete, or waiting for some other step for updated review?

**Comment by kurtmckee** [comment]
I left this open to maintain visibility of comments between it and the three PR's it is being split into.

**Comment by kurtmckee** [comment]
This work is now fully merged through smaller commits. Closing and deleting the branch.

**Pull request closed without merging**
